src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -319,18 +327,45 @@ public override Task WriteLineAsync(string value)
         /// <param name="disposing">True if called from Dispose; false if called from the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
+
+                this.textWriter?.Dispose();
             }
 
-            this.textWriter.Dispose();
+            this.disposed = true;

```

Comment:
I think we should also set `this.textWriter = null;`?

Reply:
Fixed